### PR TITLE
LifecycleTemplate; improved typing on Lifecycles in general

### PR DESCRIPTION
**Pull request by techsmyth**
API updates
- New entity: LifecycleTemplate
- Updated TemplatesSet to hold an array of LifecycleTemplates
- New scalar type to represent LifecycleDefinitions
- New privilege type: UPDATE_LIFECYCLE, currently only assigned to global admins.
- New mutation to allow updating of the LifecycleDefinition on a Challenge
- New mutation to allow updating of the LifecycleDefinition on an Opportunity.

Logic updates
- TemplateInfo.visual is no longer eager loaded: this was triggering the TypeORM issue with too many terms
- New internal server type for LifecycleDefinitions

TemplatesSet now has two default LifecycleTemplates.

Migration in place up and down, including default LifecycleTemplates.

Sample mutations:
- adding LifecycleTemplate
- updating lifecycle on Challenge
- updating lifecycle on Opportunity

Notes: 
- this PR should not require updates to the client beyond updating the schema via codegen, so existing client should just work
- core logic related to usage of Lifecycles by challenges on instantiation is not touched. 

In summary: this PR brings LifecycleTemplates into the domain model, stronger typing around LifecycleDefinitions and allows the updating of Lifecycles on Challenges + Opportunities.

### CODE REVIEW

File: src/common/interfaces/lifecycle.definition.interface.ts [hunk 1]
```diff
@@ -0,0 +1,6 @@
+export interface ILifecycleDefinition {
+  id: any;
+  context: any;
+  initial: any;
+  states: any;
+}
```
Comment (hero101) [inline src/common/interfaces/lifecycle.definition.interface.ts:6]: Can't we do something about those types? What data is in there?

File: src/domain/common/lifecycle/lifecycle.service.ts [hunk 2]
```diff
@@ -168,12 +175,24 @@ export class LifecycleService {
     return next || [];
   }
 
-  getMachineDefinition(lifecycle: ILifecycle): any {
-    return JSON.parse(lifecycle.machineDef);
+  getMachineDefinition(lifecycle: ILifecycle): ILifecycleDefinition {
+    return this.deserializeLifecycleDefinition(lifecycle.machineDef);
+  }
+
+  deserializeLifecycleDefinition(value: string): ILifecycleDefinition {
```
Comment (hero101) [inline src/domain/common/lifecycle/lifecycle.service.ts:182]: util function

File: src/domain/common/lifecycle/lifecycle.service.ts [hunk 3]
```diff
@@ -168,12 +175,24 @@ export class LifecycleService {
...
+    const result: ILifecycleDefinition = JSON.parse(value);
+    return result;
+  }
+
+  serializeLifecycleDefinition(definition: ILifecycleDefinition): string {
```
Comment (hero101) [inline src/domain/common/lifecycle/lifecycle.service.ts:187]: util function

File: src/domain/template/lifecycle-template/dto/lifecycle.template.dto.create.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    description: 'The type of the Lifecycles that this Template supports.',
+  })
+  @MaxLength(SMALL_TEXT_LENGTH)
+  type!: string;
+
+  @Field(() => LifecycleDefinitionScalar, {
+    nullable: true,
+    description: 'The XState definition for this LifecycleTemplate.',
+  })
+  @IsOptional()
```
Comment (hero101) [inline src/domain/template/lifecycle-template/dto/lifecycle.template.dto.create.ts]: Why optional?

File: src/domain/challenge/challenge/challenge.resolver.mutations.ts [hunk 5]
```diff
@@ -98,7 +99,28 @@ export class ChallengeResolverMutations {
...
   @Mutation(() => IChallenge, {
-    description: 'Updates the specified Challenge.',
+    description: 'Updates the Lifecycle on the specified Challenge.',
+  })
+  @Profiling.api
+  async updateChallengeLifecycle(
+    @CurrentUser() agentInfo: AgentInfo,
+    @Args('challengeData') challengeData: UpdateChallengeLifecycleInput
+  ): Promise<IChallenge> {
+    const challenge = await this.challengeService.getChallengeOrFail(
```
Comment (hero101) [inline src/domain/challenge/challenge/challenge.resolver.mutations.ts:109]: just the authorization is needed here

File: src/domain/challenge/challenge/challenge.service.ts [hunk 6]
```diff
@@ -176,10 +177,29 @@ export class ChallengeService {
...
       }
     }
-
     return challenge;
   }
 
+  async updateChallengeLifecycle(
+    challengeData: UpdateChallengeLifecycleInput
+  ): Promise<IChallenge> {
+    const challenge = await this.getChallengeOrFail(challengeData.challengeID);
```
Comment (hero101) [inline src/domain/challenge/challenge/challenge.service.ts]: you just need the id here

File: src/domain/challenge/challenge/challenge.service.ts [hunk 7]
```diff
@@ -176,10 +177,29 @@ export class ChallengeService {
...
+
+    const lifecycleDefinition =
+      this.lifecycleService.deserializeLifecycleDefinition(
+        challengeData.lifecycleDefinition
+      );
+    challenge.lifecycle = await this.lifecycleService.createLifecycle(
+      challenge.id,
+      lifecycleDefinition
+    );
+    await this.lifecycleService.deleteLifecycle(oldLifecycle.id);
```
Comment (hero101) [inline src/domain/challenge/challenge/challenge.service.ts]: this method impl can be changed to use the delete method of the repo, by just providing the id. Fetching the entity as in the current impl is not required.

File: src/domain/collaboration/opportunity/opportunity.service.ts [hunk 8]
```diff
@@ -111,6 +112,28 @@ export class OpportunityService {
...
+    const lifecycleDefinition =
+      this.lifecycleService.deserializeLifecycleDefinition(
+        opportunityData.lifecycleDefinition
+      );
+    opportunity.lifecycle = await this.lifecycleService.createLifecycle(
+      opportunity.id,
+      lifecycleDefinition
+    );
+    await this.lifecycleService.deleteLifecycle(oldLifecycle.id);
+    return await this.save(opportunity);
```
Comment (hero101) [inline src/domain/collaboration/opportunity/opportunity.service.ts:133]: Create, assign new lifecycle, delete old lifecycle vs just updating the existing one?

File: src/domain/template/lifecycle-template/lifecycle.template.service.ts [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    return await this.lifecycleTemplateRepository.save(lifecycleTemplate);
+  }
+
+  async deleteLifecycleTemplate(
+    lifecycleTemplate: ILifecycleTemplate
+  ): Promise<ILifecycleTemplate> {
+    const templateId: string = lifecycleTemplate.id;
+    await this.templateBaseService.deleteEntities(lifecycleTemplate);
```
Comment (hero101) [inline src/domain/template/lifecycle-template/lifecycle.template.service.ts:73]: cascade on delete vs `deleteEntities`

**Review by hero101 — APPROVED** [review]

File: src/domain/template/template-info/template.info.service.ts [hunk 10]
```diff
@@ -38,6 +42,22 @@ export class TemplateInfoService {
     return await this.templateInfoRepository.save(templateInfo);
   }
 
+  async getTemplateInfoOrFail(
+    templateInfoID: string,
+    options?: FindOneOptions<TemplateInfo>
+  ): Promise<ITemplateInfo> {
+    const templateInfo = await TemplateInfo.findOne(
```
Comment (hero101) [inline src/domain/template/template-info/template.info.service.ts]: `templateInfoRepository`

**Review by hero101 — CHANGES_REQUESTED** [review]
A complete review.
Migration tested on local.

**Review by hero101 — APPROVED** [review]

**Comment by hero101** [comment]
@Comoque1 please test this if you have the capacity.
The PR should be merged after it's tested because it introduces a lot of changes and additions.
I have reviewed and manually tested some cases.

**Review by valentinyanakiev — APPROVED** [review]
